### PR TITLE
#1-4Tasks.vladislavBirkos

### PR DESCRIPTION
**Pull request by dirtwitcher**
https://travis-ci.org/github/dirtwitcher/js-assignments/builds/711716513

### CODE REVIEW

File: task/01-strings-tasks.js [hunk 1]
```diff
@@ -69,10 +67,9 @@ function getStringFromTemplate(firstName, lastName) {
  *   'Hello, Chuck Norris!' => 'Chuck Norris'
  */
 function extractNameFromTemplate(value) {
-    throw new Error('Not implemented');
+    return value.slice(7,value.length-1);
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: `value.slice(7,-1)` will also work.
Please, rework

File: task/01-strings-tasks.js [hunk 2]
```diff
@@ -130,7 +127,7 @@ function repeatString(value, count) {
  *   'ABABAB','BA' => 'ABAB'
  */
 function removeFirstOccurrences(str, value) {
-    throw new Error('Not implemented');
+    return str.slice(0,str.indexOf(value)) + str.slice(str.indexOf(value)+value.length, str.length);
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Please, try to follow these rules https://javascript.info/coding-style  (spaces between operators particularly)

File: task/01-strings-tasks.js [hunk 3]
```diff
@@ -145,10 +142,9 @@ function removeFirstOccurrences(str, value) {
  *   '<a>' => 'a'
  */
 function unbracketTag(str) {
-    throw new Error('Not implemented');
+    return str.slice(1,str.length-1);
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Please, use `return str.slice(1,-1);`

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -236,12 +261,11 @@ function encodeToRot13(str) {
  *   isString({}) => false
  *   isString('test') => true
  *   isString(new String('test')) => true
- */
+ */ 
 function isString(value) {
-    throw new Error('Not implemented');
+       return (typeof(value) === 'string' || value instanceof String) ? true : false;
```
Comment (vramaniuk) [inline task/01-strings-tasks.js]: Ternary is redundant here
Please, just return `(typeof(value) === 'string' || value instanceof String)`

Please, see https://javascript.info/logical-operators

File: task/01-strings-tasks.js [hunk 5]
```diff
@@ -267,7 +291,17 @@ function isString(value) {
  *   'K♠' => 51
  */
 function getCardId(value) {
-    throw new Error('Not implemented');
+    let elem = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];
+    switch(value.slice(value.length-1)){
+    case '♣':
```
Comment (vramaniuk) [inline task/01-strings-tasks.js:296]: Please, improve slice second argument

File: task/02-numbers-tasks.js [hunk 6]
```diff
@@ -54,7 +53,7 @@ function getCicleCircumference(radius) {
  *  -3, 3  => 0
  */
 function getAverage(value1, value2) {
-    throw new Error('Not implemented');
+    return value1/2 + value2/2;
```
Comment (vramaniuk) [inline task/02-numbers-tasks.js]: Please, spaces around orepators

File: task/02-numbers-tasks.js [hunk 7]
```diff
@@ -73,7 +72,7 @@ function getAverage(value1, value2) {
  *   (-5,0) (10,-10) => 18.027756377319946
  */
 function getDistanceBetweenPoints(x1, y1, x2, y2) {
-    throw new Error('Not implemented');
+    return Math.sqrt((x2-x1)**2+(y2-y1)**2);
```
Comment (vramaniuk) [inline task/02-numbers-tasks.js]: Please, improve code using Math.hypot

File: task/02-numbers-tasks.js [hunk 8]
```diff
@@ -160,7 +160,7 @@ function parseNumberFromString(value) {
  *   1,2,3   => 3.741657386773941
  */
 function getParallelipidedDiagonal(a,b,c) {
-    throw new Error('Not implemented');
+    return Math.sqrt(a**2 + b**2 + c**2);
```
Comment (vramaniuk) [inline task/02-numbers-tasks.js]: Please, improve code using Math.hypot

File: task/04-arrays-tasks.js [hunk 9]
```diff
@@ -70,7 +73,9 @@ function doubleArray(arr) {
  *    [] => [] 
  */
 function getArrayOfPositives(arr) {
-   throw new Error('Not implemented');
+   return arr.filter((el)=>{
+      if (el>0) return el;
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: Please, spaces between operators

File: task/04-arrays-tasks.js [hunk 10]
```diff
@@ -85,7 +90,11 @@ function getArrayOfPositives(arr) {
  *    [ 'cat, 'dog', 'raccon' ] => [ 'cat', 'dog', 'racoon' ]
  */
 function getArrayOfStrings(arr) {
-   throw new Error('Not implemented');
+   return arr.filter((el) => {
+      if (typeof el === 'string'){
+         return el;
+      }
+   })
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: `return arr.filter((el) => {
      return (typeof el === 'string');
  })` will also work

Please, see https://javascript.info/logical-operators

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

File: task/04-arrays-tasks.js [hunk 11]
```diff
@@ -102,7 +111,11 @@ function getArrayOfStrings(arr) {
  *    [ false, 0, NaN, '', undefined ]   => [ ]
  */
 function removeFalsyValues(arr) {
-   throw new Error('Not implemented');
+   return arr.filter((el) => {
+      if (el){
+         return el;
+      }
+   })
 }
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: `arr.filter((x) => x)` will also work
Please, improve our filter callback

File: task/04-arrays-tasks.js [hunk 12]
```diff
@@ -146,7 +162,7 @@ function getStringsLength(arr) {
  *    [ 1, 'b', 'c'], 0, 'x'  => [ 'x', 1, 'b', 'c' ]
  */
 function insertItem(arr, item, index) {
-   throw new Error('Not implemented');
+   return arr.splice(index,null,item);
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: Please, code style
Space between parameters

File: task/04-arrays-tasks.js [hunk 13]
```diff
@@ -175,10 +190,9 @@ function getHead(arr, n) {
  *    [ 'a', 'b', 'c', 'd'], 3  => [ 'b', 'c', 'd' ]
  */
 function getTail(arr, n) {
-   throw new Error('Not implemented');
+   return arr.slice(arr.length-n, arr.length);
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: Spaces between operators

File: task/04-arrays-tasks.js [hunk 14]
```diff
@@ -215,10 +234,11 @@ function toCsvText(arr) {
  *   [ 10, 100, -1 ]      => [ 100, 10000, 1 ]
  */
 function toArrayOfSquares(arr) {
-   throw new Error('Not implemented');
+   return arr.map((el)=>{
+      return el*el;
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: Spaces between operators

File: task/04-arrays-tasks.js [hunk 15]
```diff
@@ -249,10 +277,15 @@ function getMovingSum(arr) {
  * [ "a" ] => []
  */
 function getSecondItems(arr) {
-   throw new Error('Not implemented');
+   let newArr = new Array();
+   arr.filter((el,index)=>{
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: Spaces between arguments

File: task/04-arrays-tasks.js [hunk 16]
```diff
@@ -268,10 +301,15 @@ function getSecondItems(arr) {
  *  [ 1,2,3,4,5 ] => [ 1, 2,2, 3,3,3, 4,4,4,4, 5,5,5,5,5 ]
  */
 function propagateItemsByPositionIndex(arr) {
-   throw new Error('Not implemented');
+   let newArr = new Array();
+   arr.map((el,index)=>{
+      let bufArr = new Array(index+1);
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: Spaces between operators

File: task/04-arrays-tasks.js [hunk 17]
```diff
@@ -286,10 +324,13 @@ function propagateItemsByPositionIndex(arr) {
  *   [ 10, 10, 10, 10 ] => [ 10, 10, 10 ]
  */
 function get3TopItems(arr) {
-   throw new Error('Not implemented');
+   arr.reverse();
+   if (arr.length<=3) {
+      return arr;
+   } 
+   return arr.slice(0, 3);
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: It won't work if initial array isn't sorted, please add sort method to solution

File: task/04-arrays-tasks.js [hunk 18]
```diff
@@ -304,7 +345,11 @@ function get3TopItems(arr) {
  *   [ 1, '2' ] => 1
  */
 function getPositivesCount(arr) {
-   throw new Error('Not implemented');
+   let count = 0;
+   arr.map((el)=>{
+      if (el>0 && typeof el === 'number') count++;
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: spaces between operators

File: task/04-arrays-tasks.js [hunk 19]
```diff
@@ -337,7 +385,9 @@ function sortDigitNamesByNumericOrder(arr) {
  *   [ 1, 10, 100, 1000 ]  => 1111
  */
 function getItemsSum(arr) {
-   throw new Error('Not implemented');
+   return arr.reduce((accum,el)=>{
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: code style

File: task/04-arrays-tasks.js [hunk 20]
```diff
@@ -353,7 +403,11 @@ function getItemsSum(arr) {
  *  [ null, undefined, NaN, false, 0, '' ]  => 6
  */
 function getFalsyValuesCount(arr) {
-   throw new Error('Not implemented');
+   let count = 0;
+   arr.map((el)=>{
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: code style

File: task/04-arrays-tasks.js [hunk 21]
```diff
@@ -371,7 +425,11 @@ function getFalsyValuesCount(arr) {
  *    [ true, 0, 1, 'true' ], true => 1
  */
 function findAllOccurences(arr, item) {
-   throw new Error('Not implemented');
+   let count = 0;
+   arr.map((el)=>{
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: code style

File: task/04-arrays-tasks.js [hunk 22]
```diff
@@ -454,7 +526,13 @@ function getIdentityMatrix(n) {
  *     3, 3   => [ 3 ]
  */
 function getIntervalArray(start, end) {
-   throw new Error('Not implemented');
+   let arr = Array(end-start+1);
```
Comment (vramaniuk) [inline task/04-arrays-tasks.js]: code style

**Review by vramaniuk — COMMENTED** [review]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by vramaniuk** [comment]
> https://travis-ci.org/github/dirtwitcher/js-assignments/builds/711716513
Please, privide the direct link to your task success travis build, when you are done with the notes

**Comment by dirtwitcher** [comment]
https://travis-ci.org/github/dirtwitcher/js-assignments/builds/721622634

**Comment by vramaniuk** [comment]
Closing the PR as accepted! Thanks!

**Pull request closed without merging**